### ISSUE TITLE
[FR] Schema & Graph Validators

### ISSUE THREAD

**Issue by paullewis** [breadboard-ai/breadboard#3321]
Related to #3266

It feels like we have a few places where we could use validators, particularly around the new Streamlined Schema Editor, where we will want to assert that the schema received is valid (and not just random JSON). But we also need to validate things like pasted graphs as well.

Should we add these to `@google-labs/breadboard`?

**Comment by dglazkov** [comment]
@aomarks I am curious if you have any machinery we can reuse here?

**Comment by aomarks** [comment]
We're using ajv in some places, so I'd use https://ajv.js.org/api.html#ajv-validateschema-schema-object-boolean to validate a JSON schema, I think.

**Comment by dglazkov** [comment]
Sound good! What about port spec schemas? Would there be something I can reuse?

**Comment by dglazkov** [comment]
Found https://www.npmjs.com/package/codemirror-json-schema

**Comment by dglazkov** [comment]
Ok, I have an idea but it might take a little while to lift/shift:

- Use ajv's [standalone validation code](https://ajv.js.org/standalone.html#generating-functions-s-for-multiple-schemas-using-the-js-library-es6-and-esm-exports) generators to create validators for all interesting types.
- To do this, we'll take @aomarks's cool `build-code` package and split it up a bit more, extracting the bits that generate schemas from TS.
- This new package, which I really want to call `@breadboard-ai/schematato`, will offer a CLI and an export that could be used in a couple of different ways.
- One is to generate schemas for `build-code`
- The other is to generate TS code for schema validation
- Then, we'll make `schematato` a dev dependency for the `breadboard` package and as part of the build process, generate validators for all the interesting types, from `Schema` to `LLMContent`.
- We could also reuse this machinery to generate `schema/breadboard.schema.json`, right?

WDYT?

**Comment by dglazkov** [comment]
Ok. maybe not. I just generated a function to validate `Schema` and it's 34K (800+ lines) of JS. Hmmm....

**Comment by dglazkov** [comment]
So maybe the easiest thing is to follow @aomarks's advice and generate schema in the `breadboard` package and then call `validateSchema` from elsewhere?

I am trying to avoid adding `ajv` as a prod dependency in breadboard. Really want to keep that package as lean as possible.

**Comment by paullewis** [comment]
I can think of a couple of options:

1. We have a separate validator package so we don't automatically get it by importing `@google-labs/breadboard`
2. We have a lightweight "this looks good enough" checker for our schema rather than using a generic validator

wdyt?

**Comment by aomarks** [comment]
> I am trying to avoid adding `ajv` as a prod dependency in breadboard. Really want to keep that package as lean as possible.

Oh, just saw this. How much does ajv increase our package size?

If it's very large, maybe we can look for another one.

I'm pretty sure we'll need a general purpose validator in our codebase, because we need to be able to do JSON schema validation dynamically, because users should be able to define their own types. Also, the size of all possible generated validators would outweigh the size of one general purpose one over time (unless we truly stick with a tiny number of possible types, which seems very limiting).

**Comment by dglazkov** [comment]
> I'm pretty sure we'll need a general purpose validator in our codebase, because we need to be able to do JSON schema validation dynamically, because users should be able to define their own types

Yeah, I had another thought this morning: we could invert the relationship between `@breadboard-ai/schema` and `@breadboard-ai/breadboard`.

Currently, the only reason `breadboard` depends on `schema` is because `schema` contains the `graph.ts` type definitions.

If instead, we made `schema` a generic validation package that other packages could use as a helper with offline (generate TS->JSON Schema bits) or online (validate schema) validation, we would have a much more flexible system.

WDYT?

**Comment by aomarks** [comment]
The original reason for creating the `schema` package was as part of the re-organization I proposed in https://github.com/breadboard-ai/breadboard/issues/539. So, it was the root of the tree -- a tiny package which just provides the core type definitions that other packages need (e.g. so that the Build API could know the GraphDescriptor definition without depending on any Breadboard runtime or unnecessary third-party dependencies).

Maybe that should actually be another package called "types" or something, so that we can use "schema" as the home for schema utilities, which is kinda the obvious right name for that stuff.

(I'm not super attached to any particular package design, though I would just suggest a general principle of trying to avoid unnecessary third-party dependencies. E.g. the build package doesn't need ajv in its deps, but the breadboard runtime and/or the visual editor maybe does.)
